service/worker/scanner/executions/shard_validator.go

```diff
@@ -0,0 +1,326 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package executions
+
+import (
+	"context"
+
+	"github.com/gocql/gocql"
+	commonpb "go.temporal.io/api/common/v1"
+	enumspb "go.temporal.io/api/enums/v1"
+	historypb "go.temporal.io/api/history/v1"
+
+	"go.temporal.io/api/serviceerror"
+
+	persistencespb "go.temporal.io/server/api/persistence/v1"
+	"go.temporal.io/server/common"
+	"go.temporal.io/server/common/persistence"
+	"go.temporal.io/server/common/persistence/serialization"
+	"go.temporal.io/server/common/persistence/versionhistory"
+	"go.temporal.io/server/common/quotas"
+)
+
+type (
+	// ShardValidator Runs check on a single shard of executions table
+	ShardValidator interface {
+		validate(shardID int32) *ShardValidationResult
+	}
+
+	shardValidatorImpl struct {
+		execMgrFactory     persistence.ExecutionManagerFactory
+		limiter            quotas.RateLimiter
+		executionsPageSize int
+		historyDB          persistence.HistoryManager
+		validators         []ExecutionValidator
+	}
+
+	// ShardValidationResult summary of shard validation
+	ShardValidationResult struct {
+		ShardID           int32
+		TotalDBRequests   int64
+		ScanStats         shardValidationReportExecutionsScanned
+		IsFailure         bool
+		Details           string
+		Error             error
+		ExecutionFailures []ExecutionValidationFailure
+	}
+
+	// ExecutionValidationFailure details of single execution validation failure.
+	ExecutionValidationFailure struct {
+		Namespace   string
+		WorkflowID  string
+		RunID       string
+		BranchID    string
+		TreeID      string
+		NextEventID int64
+		CloseStatus enumspb.WorkflowExecutionStatus
+
+		// ValidatorTag -> ValidatorFailure
+		Failures map[string]ExecutionValidatorResult
+	}
+)
+
+// CreateShardValidator returns instance of ShardValidator implementation
+func CreateShardValidator(
+	execMgrProvider persistence.ExecutionManagerFactory, limiter quotas.RateLimiter, executionsPageSize int,
+	historyDB persistence.HistoryManager, validators []ExecutionValidator,
+) ShardValidator {
+	return &shardValidatorImpl{
+		execMgrFactory: execMgrProvider, limiter: limiter, executionsPageSize: executionsPageSize,
+		historyDB: historyDB, validators: validators,
+	}
+}
+
+func (s *shardValidatorImpl) validate(shardID int32) *ShardValidationResult {
+	report := &ShardValidationResult{
+		ShardID: shardID,
+	}
+
+	execMgr, err := s.execMgrFactory.NewExecutionManager(shardID)
+	if err != nil {
+		report.IsFailure = true
+		report.Details = "Failed to receive mutableState manager"
+		report.Error = err
+		return report
+	}
+
+	var token []byte
+	isFirstIteration := true
+	for isFirstIteration || len(token) != 0 {
```

Comment:
could you use `collection.NewPagingIterator` for pagination?
ref: https://github.com/temporalio/temporal/blob/b88ecb124f347f4ab35fa8bcc334a7e1707815cf/service/history/workflowResetter.go#L530